test/Microsoft.AspNetCore.Razor.OmniSharpPlugin.Test/DefaultProjectChangePublisherTest.cs

```diff
@@ -0,0 +1,165 @@
+                {
+                    attemptedToSerialize = true;
+
+                    Assert.Same(omniSharpProjectSnapshot, snapshot);
+                    Assert.Equal(expectedPublishFilePath, publishFilePath);
+                });
+            publisher.SetPublishFilePath(omniSharpProjectSnapshot.FilePath, expectedPublishFilePath);
+
+            // Act
+            publisher.Publish(omniSharpProjectSnapshot);
+
+            // Assert
+            Assert.True(attemptedToSerialize);
+        }
+
+        [Fact]
+        public void ProjectAdded_PublishesToCorrectFilePath()
+        {
+            // Arrange
+            var snapshotManager = CreateProjectSnapshotManager(allowNotifyListeners: true);
+            var attemptedToSerialize = false;
+            var expectedPublishFilePath = "/path/to/obj/bin/Debug/project.razor.json";
+            var publisher = new TestProjectChangePublisher(
+                LoggerFactory,
+                onSerializeToFile: (snapshot, publishFilePath) =>
+                {
+                    attemptedToSerialize = true;
+
+                    Assert.Equal(expectedPublishFilePath, publishFilePath);
+                });
+            publisher.Initialize(snapshotManager);
+            var hostProject = new OmniSharpHostProject("/path/to/project.csproj", RazorConfiguration.Default);
+            publisher.SetPublishFilePath(hostProject.FilePath, expectedPublishFilePath);
+
+            // Act
+            snapshotManager.ProjectAdded(hostProject);
+
+            // Assert
+            Assert.True(attemptedToSerialize);
+        }
+
+        [Fact]
+        public void ProjectChanged_PublishesToCorrectFilePath()
+        {
+            // Arrange
+            var snapshotManager = CreateProjectSnapshotManager(allowNotifyListeners: true);
+            var attemptsToSerialize = 0;
+            var expectedPublishFilePath = "/path/to/obj/bin/Debug/project.razor.json";
+            var publisher = new TestProjectChangePublisher(
+                LoggerFactory,
+                onSerializeToFile: (snapshot, publishFilePath) =>
+                {
+                    attemptsToSerialize++;
+
+                    Assert.Equal(expectedPublishFilePath, publishFilePath);
+                });
+            publisher.Initialize(snapshotManager);
+            var hostProject = new OmniSharpHostProject("/path/to/project.csproj", RazorConfiguration.Default);
+            publisher.SetPublishFilePath(hostProject.FilePath, expectedPublishFilePath);
+            snapshotManager.ProjectAdded(hostProject);
+            var newConfiguration = RazorConfiguration.Create(RazorLanguageVersion.Experimental, "Custom", Enumerable.Empty<RazorExtension>());
+            var newHostProject = new OmniSharpHostProject(hostProject.FilePath, newConfiguration);
+
+            // Act
+            snapshotManager.ProjectConfigurationChanged(newHostProject);
+
+            // Assert
+            Assert.Equal(2, attemptsToSerialize);
```

Comment:
I don't understand what's going on here. Why is this attempted twice?

Reply:
1 for the added project 1 for the changed project.

Reply:
It would be great to make the `ProjectAdded` an `//Act` and also rename this test to be more descriptive.

Reply:
ProjectAdded is covered in another test.